### ISSUE TITLE
Consider writing a daemon example as an actual daemon, not a web application

### ISSUE THREAD

**Issue by oising** [Azure-Samples/ms-identity-aspnet-daemon-webapp#10]
Why is every Microsoft example of a daemon, non-interactive graph client implemented as a web application? This makes zero sense and only serves to confuse the basic concepts far more than necessary.

Every time I find an example here, it's an MVC app, or a Web API app. This is **NOT** what the real world wants to see when you talk about daemon sample code. A daemon should be at a minimum implemented as a headless, non-interactive console application. This is then easily shimmed into a Azure Web Job, Azure Function or a Windows Service. Trying to remove the cruft around a web application and get one's head around how admin consent is applied to an app registration destined for a console app is far more muddy than it needs to be.

**Comment by jmprieur** [comment]
Hello @oising 
Thanks for the feedback, and actually I completely agree myself that illustrating the client credentials flow with a Web App, even if this is an interesting scenario, is overkill to learn. 

However, we have actually 3 .NET samples of daemons which are console applications. See:
https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-code-samples#daemon-applications-accessing-web-apis-with-the-applications-identity

Now if you tell me that you'd want a daemon application requesting tokens for the V2 endpoint and you'd like this sample to be a desktop app, I'm with you, and that's on the backog. Is it your case? how urgent would it be?

**Comment by oising** [comment]
"Now if you tell me that you'd want a daemon application requesting tokens for the V2 endpoint and you'd like this sample to be a desktop app"

Not a desktop app (no WPF or extraneous layering please) -- just a bare bones console app, with instructional on how to set up the app registration in the portal for a v2 endpoint, and also - importantly - how admin consent is granted BY SOMEONE ELSE for extended application permissions, like "user.read.all" via the portal, involving someone with the appropriate permissions  -- don't get lazy by avoiding this, please :) So, to summarise - cover the most common case/workflow like this:

- a bare bones console app
- uses application identity (e.g. MSI) when deployed, developer identity when testing locally
- app registration via apps.dev.microsoft.com (is this preferred over portal.azure.com? please clarify)
- uses v2 endpoint
- uses latest graph API library, not older AD graph API
- requires admin consent for "user.read.all" permission -- e.g. full directory profile
- assume the developer is NOT admin, so cannot grant permissions himself/herself, ergo someone more privileged has to do this step - make this step distinct as part of the demo please

It's not urgent per-se, because I already wasted^H^H^H^H^H^H spent two days figuring out how it needs to be done, and as such it's not a priority for me. 

Hope this helps.

**Comment by jmprieur** [comment]
Sorry @oising : I was not clear. I mean a console app (which runs on desktop)
Thanks very much for the spec of the sample. I'll definitively do that.

If you want the app to use MSI, maybe @skwan you have a sample already?

**Comment by oising** [comment]
@jmprieur Thank you. I do understand how to use MSI, but the point is there are different samples showing different aspects of integration. It would be nice to have all in one. We had to merge three different samples and it wasn't easy to figure it out due to all of the extraneous cruft :) Thank you for your responsiveness here.

**Comment by oising** [comment]
Essentially a daemon should be modeled as service-to-service. There is some good information here, with respect to dealing with key vault. So instead of key vault, use graph API as the endpoint: https://docs.microsoft.com/en-us/azure/key-vault/service-to-service-authentication

**Comment by omora** [comment]
I miss an example to access O365 sharepoint site (using graph api) from a daemon (console app on premise) How can this daemon app get access granted if it is not an URL? Anyone have an example?

**Comment by jmprieur** [comment]
@oising @darrelmiller @TheConservativeCoder @Brandon-combes @omora 

We've created a much simpler console daemon application, see the following sample: [active-directory-dotnetcore-daemon-v2](https://github.com/Azure-Samples/active-directory-dotnetcore-daemon-v2)

A sample using MSI and KeyVault is still needed
We are also thinking of adding support for managed identity in MSAL in the future

**Comment by oising** [comment]
@jmprieur Good job so far. It definitely goes a long way towards unblocking people who get blocked when they see they have to register their console app as a "web" application and provide an URL endpoint. That said, there's got to be some room for helpful inline text on the portal itself. It's confusing as hell, especially when you see that prior to this new sample, all of the other examples _were_ actually web apps, which makes me think that MS devs were just as confused as the rest of us.

**Comment by racingcow** [comment]
@jmprieur,

[active-directory-dotnetcore-daemon-v2](https://github.com/Azure-Samples/active-directory-dotnetcore-daemon-v2) seems to only work inside its own tenant. Is there a sample of a "real" daemon (not web app) that is multi-tenant like ms-identity-aspnet-daemon-webapp?

I'm trying to make a daemon that can use the graph api to read office 365 email in other tenants.

**Comment by jmprieur** [comment]
@racingcow : if in the app registration step (https://github.com/Azure-Samples/active-directory-dotnetcore-daemon-v2#register-the-client-app-daemon-console) you choose "in any organizarion", then it will be multi-tenant.
But of course the tenant admin of each tenants need to provision a service principal and approve it.

**Comment by jmprieur** [comment]
Closing this issue as the following [PR](https://github.com/Azure-Samples/active-directory-dotnetcore-daemon-v2/pull/32) is bringing more towards the goal (including a pointer to an MSI sample). Will track the MSI/KeyVault requirement in https://github.com/Azure-Samples/active-directory-dotnetcore-daemon-v2

**Comment by fahadashiq12** [comment]
This questions comes in the AZ-204 exam as well.
